client/landing/stepper/README.md

```diff
@@ -1,106 +1,387 @@
+	</ol>
+</p>
+</details>
+
+In general, the smarter the step, the more problematic and tailored it is. Please treat your steps as buttons or inputs.
+
+### Making a flow
+
+A flow is a collection of steps. Each of these steps submit some information to the flow, which means the state of the flow is largely the sum of these submitted data. For that reason, the `FlowV2` interface requires the steps collection to be defined before the flow itself. This way, the flow can shape its state around the submissions and properties of these steps.
+
+**Note:** We have an example flow you can use as a reference [here](/client/landing/stepper/declarative-flow/flows/00-example-flow/example.ts).
+
+#### Code example
 
 ```ts
-import type { StepPath } from './internals/steps-repository';
-import type { Flow } from './internals/types';
+// We define our steps collection upfront because it is a prerequisite of shaping the flow.
+async function initialize( calypsoReduxStore: Store ) {
+	const includeDomainsStep = await isTheMoonFull();
+	const hasAnySites = userHasAnySites( calypsoReduxStore.getState() );
 
-export const exampleFlow: Flow = {
-	initialize() {
-		return [ STEPS.DOMAINS, STEPS.DESIGN ];
-	},
-	useStepNavigation( currentStep, navigate ) {
-		const goBack = () => {
-			if ( currentStep === 'domain' ) {
-				navigate( 'design' );
-			} else {
-				navigate( 'domain' );
+	if ( includeDomainsStep ) {
+		return [ STEPS.UNIFIED_DOMAINS, STEPS.UNIFIED_PLANS, STEPS.PROCESSING ] as const;
+	}
+
+	if ( hasAnySites ) {
+		return [ STEPS.PICK_SITE, STEPS.UNIFIED_DOMAINS, STEPS.UNIFIED_PLANS, STEPS.PROCESSING ] as const;
+	}
+
+	// We need `as const` to promise TS that these steps won't change later.
+	return [ STEPS.UNIFIED_PLANS, STEPS.PROCESSING ] as const;
+}
+
+export const exampleFlow: FlowV2< typeof initialize > = {
+	/**
+	 * The name of the flow is what appears in the pathname. It must be unique.
+	 * This flow will be under /setup/my-flow.
+	 */
+	name: 'my-flow',
+
+	/**
+	 * This flag must be `true` for signup flows (generally where a new site may be created), and should be `false` for other flows.
+	 * It controls whether we'll trigger a `calypso_signup_start` Tracks event when the flow starts.
+	 */
+	isSignupFlow: true,
+	initialize,
+	/**
+	 * This hook is the control unit of your flow. It is where:
+	 * 1. You handle and use the information submitted by the steps.
+	 * 2. You decide to navigate to which steps when.
+	 * 3. You manage the state of the flow.
+	 */
+	useStepNavigation( currentStepSlug, navigate ) {
+		// This hook should be enough to manage and persist all the state your flows needs.
+		const { get, set } = useFlowState();
+
+		// This function handles the submission event of every step in the flow.
+		const submit: SubmitHandler< typeof initialize > = async ( submittedStep ) => {
+			const { slug, providedDependencies } = submittedStep;
+			switch ( slug ) {
+				case 'domains': {
```

Comment:
Any reasons for using a plain text string instead of, say `STEPS.DOMAINS.slug`?

Also a curiosity: is it expected that some steps have the same slug?
E.g. [`DOMAINS` and `DOMAIN_TRANSFER_DOMAINS`](https://github.com/Automattic/wp-calypso/blob/6855f70239798896d085a3049fbb0f0fec352f62/client/landing/stepper/declarative-flow/internals/steps.tsx#L54-L62)

Reply:
> Any reasons for using a plain text string instead of, say STEPS.DOMAINS.slug?

With Stepper 2.0, the `slug` is typed to the exact slugs of the steps in the flow at hand. So if you make a typo, TS will catch it. Which makes `domains` safe and less verbose than `STEPS.DOMAINS.slug`

> Also a curiosity: is it expected that some steps have the same slug?

Not really. But historically, flows made and inlined their own step objects, as in `{ slug: X, asyncComponent: Y }`. So we had all kinds of slugs, including some clashing ones. When we unified all the steps into a canonical dictionary, the clashes became apparent. I'm working on making slugs unique [here](https://github.com/Automattic/wp-calypso/issues/100585).